### PR TITLE
feature: [Table] Pagination on the top in addition to the bottom. fix…

### PR DESCRIPTION
**Pull request by kanweiwei**
new feature related (#9346)

* [x] Make sure that you propose PR to right branch: bugfix for `master`, feature for latest active branch `feature-x.x`.
* [x] Make sure that you follow antd's [code convention](https://github.com/ant-design/ant-design/wiki/Code-convention-for-antd).
* [x] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [x] Rebase before creating a PR to keep commit history clear.
* [x] Add some descriptions and refer relative issues for you PR.

Extra checklist:
 *isNewFeature* **:**

  * [x] Update API docs for the component.
  * [x] Update TypeScript definition for the component.
  * [x] Add unit tests for the feature.

### CODE REVIEW

File: components/table/Table.tsx [hunk 1]
```diff
@@ -983,6 +983,29 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
...
+          ];
+        } else {
+          return [
+            table,
+            this.renderPagination(),
+          ];
+        }
+      }
+      return table;
+    };
```
Comment (yesmeck) [inline components/table/Table.tsx]: Could you refactor it to less `if`s? Like:

```javascript
{this.renderPagination('top')}
{table}
{this.renderPagination('bottom')}
```

Comment (kanweiwei) [reply components/table/Table.tsx]: good idea, thanks

**Comment by kanweiwei** [comment]
@yesmeck  why it failed, can you help me

File: components/pagination/index.en-US.md [hunk 2]
```diff
@@ -27,6 +27,7 @@ A long list can be divided into several pages by `Pagination`, and only one page
 | itemRender | to customize item innerHTML | (page, type: 'page' \| 'prev' \| 'next', originalElement) => React.ReactNode | - |
 | pageSize | number of data items per page | number | - |
 | pageSizeOptions | specify the sizeChanger options | string\[] | ['10', '20', '30', '40'] |
+| position | specify the position of `Pagination` component when using `Table` component | 'top' \| 'bottom' \| 'both' | - |
```
Comment (yesmeck) [inline components/pagination/index.en-US.md]: Should add to `Table`.

File: components/table/demo/dynamic-settings.md [hunk 3]
```diff
@@ -70,7 +70,7 @@ class Demo extends React.Component {
   state = {
     bordered: false,
     loading: false,
-    pagination: true,
+    pagination: { position: 'top' },
```
Comment (yesmeck) [inline components/table/demo/dynamic-settings.md]: Do not need to change this demo.

File: components/pagination/index.en-US.md (hunk 2 above)
Comment (yesmeck) [reply components/pagination/index.en-US.md]: Like `rowSelection` does.

**Comment by yesmeck** [comment]
Add a test case, please.

File: components/table/index.en-US.md [hunk 4]
```diff
@@ -65,7 +65,7 @@ const columns = [{
 | indentSize | Indent size in pixels of tree data | number | 15 |
 | loading | Loading status of table | boolean\|[object](https://ant.design/components/spin-cn/#API) ([more](https://github.com/ant-design/ant-design/issues/4544#issuecomment-271533135)) | `false` |
 | locale | i18n text including filter, sort, empty text, etc | object | filterConfirm: 'Ok' <br> filterReset: 'Reset' <br> emptyText: 'No Data' <br> [Default](https://github.com/ant-design/ant-design/issues/575#issuecomment-159169511) |
-| pagination | Pagination [config](/components/pagination/), hide it by setting it to `false` | object |  |
+| pagination | Pagination [config](#pagination) or [`Pagination`] (/components/pagination/), hide it by setting it to `false` | object |  |
```
Comment (yutingzhao1991) [inline components/table/index.en-US.md]: I don't agree copy pagination api at here. link to `/components/pagination-cn/#API` is ok.

Comment (kanweiwei) [reply components/table/index.en-US.md]: but @yesmeck sugggest me to add the new feature to `Table`, I'm confused

File: components/table/index.en-US.md [hunk 5]
```diff
@@ -135,6 +135,29 @@ One of the Table `columns` prop for describing the table's columns, Column has t
...
+| Property | Description | Type | Default |
+| -------- | ----------- | ---- | ------- |
+| current | current page number | number | - |
+| defaultCurrent | default initial page number | number | 1 |
+| defaultPageSize | default number of data items per page | number | 10 |
+| hideOnSinglePage | Whether to hide pager on single page | boolean | false |
+| itemRender | to customize item innerHTML | (page, type: 'page' \| 'prev' \| 'next', originalElement) => React.ReactNode | - |
+| pageSize | number of data items per page | number | - |
+| pageSizeOptions | specify the sizeChanger options | string\[] | ['10', '20', '30', '40'] |
+| position | specify the position of `Pagination` | 'top' \| 'bottom' \| 'both' | 'bottom' |
```
Comment (yesmeck) [inline components/table/index.en-US.md]: Don't need copy all Pagination props to here. Just leave `positon` here, and link others to `/components/pagination-cn/#API`.

Comment (yesmeck) [inline components/table/index.en-US.md]: Don't need copy all Pagination props to here. Just leave `positon` here, and link others to `/components/pagination-cn/#API`.

**Review by yesmeck — CHANGES_REQUESTED** [review]

**Comment by kanweiwei** [comment]
@yesmeck @yutingzhao1991  When i run `npm run lint:demo`, it threw the error：
image
The Travis CI buid failed beacause of the error. Can you tell me how to fix it?

File: components/form/__tests__/__snapshots__/demo.test.js.snap [hunk 6]
```diff
@@ -2503,7 +2503,7 @@ exports[`renders ./components/form/demo/validate-other.md correctly 1`] = `
             />
             <div
               class="ant-slider-track"
-              style="visibility:visible;left:0%;width:0%"
+              style="left:0%;width:0%"
```
Comment (yesmeck) [inline components/form/__tests__/__snapshots__/demo.test.js.snap]: 不相关的 snapshot 不要提交

Comment (yutingzhao1991) [reply components/form/__tests__/__snapshots__/demo.test.js.snap]: 但是这样会导致测试失败，这种情况一般怎么处理？我们来更新 snapshot ?

Comment (yesmeck) [reply components/form/__tests__/__snapshots__/demo.test.js.snap]: `feature-3.3.0` 上已经更新了， rebase 一下就可以。

File: components/table/index.en-US.md (hunk 4 above)
Comment (yutingzhao1991) [reply components/table/index.en-US.md]: I guess @yesmeck means add a new props to `Table`, not add `position` to `pagination`.  ping @yesmeck .

Comment (yesmeck) [reply components/table/index.en-US.md]: 是这样 `<Table pagination={{ position: 'both' }} />`

File: components/pagination/Pagination.tsx [hunk 7]
```diff
@@ -16,6 +16,7 @@ export interface PaginationProps {
   hideOnSinglePage?: boolean;
   showSizeChanger?: boolean;
   pageSizeOptions?: string[];
+  position?: 'top' | 'bottom' | 'both';
```
Comment (yesmeck) [inline components/pagination/Pagination.tsx]: Should move to Table

Comment (kanweiwei) [reply components/pagination/Pagination.tsx]: Table初始化的时候。this.state.pagination 指定为Pagination的prop类型，所以我在Pagination的prop类型声明里加了这个属性
image

Comment (yesmeck) [reply components/pagination/Pagination.tsx]: 可以在 Table 里新增一个 `TablePaginationConfig` 类型，继承 `PaginationProps`。

**Review by yesmeck — APPROVED** [review]

**Comment by MrSaints** [comment]
Awesome :)

**Comment by JeanZhao** [comment]
Hello, could anybody know how to set the pagination position as 'top' with antd v2.x ? There are some reasons we couldn't just directly upgrade it from v2.x to v3.x...

**Comment by JeanZhao** [comment]
@yesmeck @dengfuping

**Comment by kanweiwei** [comment]
> Hello, could anybody know how to set the pagination position as 'top' with antd v2.x ? There are some reasons we couldn't just directly upgrade it from v2.x to v3.x...

you can define table component with `Table` and `Pagination` by yourself, like this:
```
<CustomTable pagination={pagination} table={tableProperties} />;

function CustomTable(props: any) {
  return (
    <div>
      <div className="header">
        <Pagination {...props.pagination} />
      </div>
      <Table {...props.tableProperties} />
    </div>
  );
}
```

**Comment by JeanZhao** [comment]
> > Hello, could anybody know how to set the pagination position as 'top' with antd v2.x ? There are some reasons we couldn't just directly upgrade it from v2.x to v3.x...
> 
> you can define table component with `Table` and `Pagination` by yourself, like this:
> 
> ```
> <CustomTable pagination={pagination} table={tableProperties} />;
> 
> function CustomTable(props: any) {
>   return (
>     <div>
>       <div className="header">
>         <Pagination {...props.pagination} />
>       </div>
>       <Table {...props.table} />
>     </div>
>   );
> }
> ```
You mean, I could use Pagination component individually with v3.x? And then mix use with antd table?

**Comment by kanweiwei** [comment]
@JeanZhao yes, mix use them in v2.x. You can have a try. In that case, you should notice that Antd `Table` accepts pagination property with `null`, then you control the pagination.

**Comment by JeanZhao** [comment]
> @JeanZhao yes, mix use them in v2.x. You can have a try. In that case, you should notice that Antd `Table` accepts pagination property with `null`, then you control the pagination.

Sorry, I'm not so clear...Pagination component in v2.x. doesn't accpet PaginationProps 'position' . How do it work when just write as what you suggest:
<Pagination  position='top' />

**Comment by JeanZhao** [comment]
额。。可以直接说中文，，我忘记了，antd 团队是国内的。。

**Comment by kanweiwei** [comment]
...你用2.x版本的antd的table和pagination组件，自己封装一个组件。你给table的pagintion属性传null，让antd 的 table不要显示分页器。自己用pagintion组件，排到表格前面还是表格后面，具体看你的需求。

**Pull request merged**